### PR TITLE
Create New Game mode - CHOOSE

### PR DESCRIPTION
**Pull request by jriyyya**
For this game mode the player has to choose from the cats displayed on the screen which has a odd count.

edit : Added another type in this game mode, which now ask the player to choose the cat on the screen which has an even count.

@sourabhaa @chimosky Please Review

### CODE REVIEW

File: gamearea.py
Comment (chimosky) [inline gamearea.py:1]: You could merge this with the commit where `__draw_choose_options` is created.

File: gamearea.py [hunk 1]
```diff
@@ -268,7 +268,7 @@ def __release_cb(self, widget, event):
         self.selected_cat = None
         if "type" not in self.level_data.keys():
             return 0
-        if self.level_data["type"] == GameType.ROWS:
+        if self.level_data["type"] == GameType.ROWS or self.level_data["type"] == GameType.CHOOSE:
```
Comment (chimosky) [inline gamearea.py:272]: The commit message says " Add functionlity of _release_cb in gameType CHOOSE" but the change in the commit doesn't reflect that.

File: gamearea.py [hunk 2]
```diff
@@ -441,8 +442,21 @@ def __draw_end_message(self, context):
...
+        elif self.level_data["type"] == GameType.CHOOSE:
+            if self.selected_option is not None:
+                if self.selected_option.cat_id == self.odd_cat_id:
+                    message = _("You Choose correctly!")
+                    self.win = True
+                else:
+                    message = _("You Choose wrong")
+                    self.win = False
+            else:
+                message = _("You should have chosen an option")
```
Comment (chimosky) [inline gamearea.py]: choose should be small letters and it should be in past tense as the time this message shows, you're expected to have made your choices.

File: gamearea.py [hunk 3]
```diff
@@ -511,6 +511,25 @@ def show_message(self, context, message, font_size, y=0):
...
+
+        for i in range(3):
+            even_number = random.randrange(0, remaining_sum - (3 - i) * 2 + 2, 2)
+            divided_nums.append(even_number)
+            remaining_sum -= even_number
+        divided_nums.append(remaining_sum)
+        divided_nums[self.odd_cat_id - 1] += 1
+
+        for i in range(0, 4):
+            for j in range(0, divided_nums[i]):
```
Comment (chimosky) [inline gamearea.py]: You can omit the 0's in this loop.

**Comment by chimosky** [comment]
Reviewed 5a7dc44, not tested.

File: gamearea.py (hunk 1 above)
Comment (jriyyya) [reply gamearea.py:272]: Yes, I realised it was quite confusing.

I have changed the commit message [7f629b2](https://github.com/sugarlabs/classify-cats/pull/9/commits/7f629b24a274f77578c2e64c7cef418a725eb857)

**Comment by jriyyya** [comment]
@chimosky I have made the suggested changes

File: gamearea.py [hunk 4]
```diff
@@ -237,6 +237,15 @@ def __motion_cb(self, widget, event):
                 self.over_option = new_option
                 self.redraw()
 
+        elif self.level_data["type"] == GameType.CHOOSE:
+            flag_hover = False
+            for cat in self.choose_option_cats:
+                if event.x > cat.x and event.x < cat.x + cat.width and event.y > cat.y and event.y < cat.y + cat.height:
```
Comment (chimosky) [inline gamearea.py]: This could be made more readable as it looks confusing.

**Review by chimosky — CHANGES_REQUESTED** [review]

File: gamearea.py (hunk 2 above)
Comment (chimosky) [reply gamearea.py]: This still uses "Choose" and not "chose".

File: gamearea.py (hunk 3 above)
Comment (chimosky) [reply gamearea.py]: Any reason why you still have the range starting from 0?

File: gamearea.py [hunk 5]
```diff
@@ -546,18 +546,24 @@ def add_cats(self):
...
 
-        for x in range(0, cats):
-            cat_id = random.choice(cat_ids)
-            cat = Cat(cat_id, cat_width, cat_height)
+        elif self.level_data["type"] == GameType.CHOOSE:
+            choose_cat_ids = self.divide_into_even_odd(cats)
+            cat_width = 60
+            cat_height = 60
 
+        for x in range(0, cats):
```
Comment (chimosky) [inline gamearea.py]: Another range starting with 0.

File: gamearea.py (hunk 2 above)
Comment (chimosky) [reply gamearea.py]: Seen your change.

File: gamearea.py (hunk 4 above)
Comment (jriyyya) [reply gamearea.py]: Done

File: gamearea.py [hunk 6]
```diff
@@ -240,7 +240,7 @@ def __motion_cb(self, widget, event):
         elif self.level_data["type"] == GameType.CHOOSE:
             flag_hover = False
             for cat in self.choose_option_cats:
-                if event.x > cat.x and event.x < cat.x + cat.width and event.y > cat.y and event.y < cat.y + cat.height:
+                if (cat.x < event.x < cat.x + cat.width) and (cat.y < event.y < cat.y + cat.height):
```
Comment (chimosky) [inline gamearea.py]: This isn't that readable, a better check would be;
```python3
if event.x > cat.x and event.x < (cat.x + cat.width) and event.y > cat.y and event.y < (cat.y + cat.height):
```

**Comment by chimosky** [comment]
Tested 863f284, works as expected although the time it takes to start the choose game mode is quite short considering the text that's displayed on the screen and sometimes it makes it a bit difficult to read the instruction for the game mode.

**Comment by jriyyya** [comment]
image


@chimosky Is it more readable now?

**Comment by chimosky** [comment]
> @chimosky Is it more readable now?

Yes it is.

I've made a small change, it'll be great to add an event count mode too, not just odd.

**Comment by jriyyya** [comment]
So I should continue this PR, or can I create a new PR for this?

**Comment by jriyyya** [comment]
So should I continue this PR, or can I create a new PR for this?

**Comment by chimosky** [comment]
> So should I continue this PR, or can I create a new PR for this?

This is for the new game mode so it makes sense to add a part of the said game mode here.

**Comment by jriyyya** [comment]
> This is for the new game mode so it makes sense to add a part of the said game mode here.

Okay, I will do so

File: gamearea.py [hunk 7]
```diff
@@ -522,7 +531,16 @@ def divide_into_even_odd(self, number):
             divided_nums.append(even_number)
             remaining_sum -= even_number
         divided_nums.append(remaining_sum)
-        divided_nums[self.odd_cat_id - 1] += 1
+        print(divided_nums)
```
Comment (chimosky) [inline gamearea.py]: I'm guessing you forgot to remove this after debugging.

File: gamearea.py [hunk 8]
```diff
@@ -522,7 +531,16 @@ def divide_into_even_odd(self, number):
...
+        if self.choose_type:
+            divided_nums[self.choose_cat_id - 1] += 1
+            
+        else:
+            for i in range(0, 4):
+                if (i + 1) != self.choose_cat_id:
+                    divided_nums[i] += 1
+
+        print(divided_nums, self.choose_type)
```
Comment (chimosky) [inline gamearea.py]: This too.

**Review by chimosky — CHANGES_REQUESTED** [review]

File: gamearea.py (hunk 7 above)
Comment (jriyyya) [reply gamearea.py]: Yes, Thank you.

Also I was trying out something before and forgot to remove some parts of it and committed it by mistake, removed them as well.

**Comment by chimosky** [comment]
I tried testing and I never encountered the odd game mode but I did notice something odd;

[Screencast from 2023-09-01 18-05-41.webm](https://github.com/sugarlabs/classify-cats/assets/16203834/820e1d12-b768-4771-91c7-1c01bd4538d3)

As you'll notice from the screencast, I fail the question after choosing correctly, I chose blue which was even and failed.
Noticing this for the first time hence bringing it up and it seems it happens when they're two cats that fall into the category of the question - not shown in the video above - .

**Comment by jriyyya** [comment]
The game gets over after some specific level count, and at the end of the screen the final score is shown, I think it added up the score to the final score, so you didn't fail the final round.

Also I was using random to choose between 0 and 1 for even and odd game mode, very strange that you weren't able to encounter the odd game mode 😅

**Comment by chimosky** [comment]
> The game gets over after some specific level count, and at the end of the screen the final score is shown, I think it added up the score to the final score, so you didn't fail the final round.

We do need a help button to explain that as it's not obvious, my experience is about the game play and not the end. 

> Also I was using random to choose between 0 and 1 for even and odd game mode, very strange that you weren't able to encounter the odd game mode 😅

How was 4d6b81c buggy?

**Comment by jriyyya** [comment]
> We do need a help button to explain that as it's not obvious, my experience is about the game play and not the end. 

I will look into it.

> How was 4d6b81c buggy?

I was thinking about merging the win check condition for game type row and choose, and later realised it is kind of not a good idea, and I forgot to remove some of it's code

**Comment by jriyyya** [comment]
Hi @chimosky, Fixed it.

**Comment by jriyyya** [comment]
@chimosky Can you please merge this, We can then make a release for this.

**Comment by chimosky** [comment]
Tested 47c8b9a, works as expected. 

We can look at making the game not end where it does, it's quite short.

Thanks for working on this!

**Comment by jriyyya** [comment]
@chimosky How about we make the max puzzle count as 10 so that the game doesn't end soon, currently it's 5

**Pull request merged**